### DISCUSSION TITLE
How to make markdown headings be left justified?

### DISCUSSION THREAD

**Discussion by ssbarnea** [Textualize/rich#2413] [Q&A]
The default behavior of rich is to render all headings in `Markdown()` as justified, and that is not really how markdown is currently rendered by most other renderers, as they prefer to just keep the title left justified instead of centered.

How can I change this?

**Comment by willmcgugan** [comment]
See [Markdown.elements](https://github.com/Textualize/rich/blob/master/rich/markdown.py)

**Reply by TechWiz-3** [reply]
So from what my beginner eyes understand line 145 basically means we can't change it to the left?
https://github.com/Textualize/rich/blob/bfea226da7baddd3c5e31c9f0c0d2f2ca45e4c7c/rich/markdown.py#L145

**Reply by ssbarnea** [reply]
I think that what @willmcgugan suggested without effectively naming it, was monkey-patching rich code in order to change its styling. As polishing the rendering was not my top priority, I did not had time to test it, especially as monkey-patching code that end-us in production is usually not the best approach.

I gave it a go and apparently it worked, obviously with the risk of messing stuff on new versions. I will post a link once I have my PR ready.

**Reply by TechWiz-3** [reply]
Cool thanks a lot for the response. I tried it on my system as well and it worked.

Do you think it would be worth submitting a PR? I think I can figure out how to make the heading justification configurable, would be good learning for me 🤔

**Reply by willmcgugan** [reply]
It's not monkey patching. I designed it so that you could customize the various blocks.

@TechWiz-3 That's not something I would accept. But I'm sure you could find plenty things to contribute to. See the issues tab...

**Reply by ssbarnea** [reply]
My WIP code to alter the rendering is at https://github.com/ansible/ansible-lint/pull/2486 but I am still trying to find a way to disable the border that is added on code snippets and to disable their custom background. By looking at what happens I expected to find a `box` somewhere and an additional one space indentation added to each line.

![](https://sbarnea.com/ss/Screen-Shot-2022-09-22-09-52-24.36.png)

**Reply by TechWiz-3** [reply]
>  I designed it so that you could customize the various blocks.

So it can be changed to `left` without changing the source code? Sorry if I'm being dense, I've read through the docs numerous times and looked at `markdown.py` and still have no idea 😅

**Reply by ssbarnea** [reply]
Sadly, with the recent change in rich 13.2, the override we used no longer works so basically we are forced to accept the "centered" decision that rich made.

It is bit annoying as almost nobody rendering markdown to html does centered headers, is only a rich thing, probably inspired from by the lynx browser.
